File: Allegro5 Test/PlayerWhite.cpp
#include "PlayerWhite.h"
PlayerWhite::PlayerWhite(){}

void PlayerWhite::Destroy(){
	GameObject::Destroy();
}

void PlayerWhite::Init(ALLEGRO_BITMAP *image = 0,int spawnX = 350, int spawnY = 100){
	GameObject::Init(spawnX, spawnY,4,4,0,1,5,32);
	SetID(PLAYERWHITE);
	SetAlive(true);
	SetCollideType(WHITE);

	this->spawnX = spawnX;
	this->spawnY = spawnY;
	lives = 100;
	score = 0;
	jump=false;
	mayJumpAgain = false;
	isOnSolidGround = false;
	willJumpLeft = willJumpRight = false;
	jumpFramesLeft = 0;
	gravity = 1;
	jumpSpeed = 15;
	float cameraPosition[2] = {0,0};
	dying = false;

	frameDelay = 5;
	frameCount = 0;
	maxFrame = 8;
	curFrame = 0;
	frameWidth = 64;
	frameHeight = 64;
	animationColumns = 8;
	animationDirection = 0;

	animationRow = 8;

	if(image != 0){
		PlayerWhite::image = image;
	}
}

void PlayerWhite::Update(){

	GameObject::Update();

	if(x<0){
		x=0;
	}
	else if(x > WIDTH)
		x = WIDTH;
	if(y<=0){
		dirY = 1;
	}
	/*
	else if(y > HEIGHT){
		if(lives>0){
			dieAnimation();
		}
	}*/	

	CameraUpdate(cameraPosition, x, y, frameWidth, frameHeight);
	frameCount++;
}

void PlayerWhite::Render(){
	GameObject::Render();

	int fx = (curFrame % animationColumns) * frameWidth;
	int fy = animationRow * frameHeight;

	al_draw_bitmap_region(image, fx, fy, frameWidth, frameHeight, 
		x - frameWidth/2, y - frameHeight/2, 0);


}

void PlayerWhite::dieAnimation(){
	dying = true;
	velX = 1;
	if(frameCount>=frameDelay){
		frameCount = 0;
		if(dirX == 1){
			if(animationRow!=2){
				animationRow = 2;
				curFrame = 0;
			}
			else if(curFrame == 7){
				LoseLife();
				dying = false;
				if(lives == 0)
					SetAlive(false);
				else{
					x = spawnX;
					y = spawnY;
					ResetAnimation(RIGHT);
					ResetMovement(RIGHT);
				}
			}
			else
				++curFrame;
		}
		else{
			if(animationRow!=11){
				animationRow = 11;
				curFrame = 7;
			}
			else if(curFrame == 0){
				LoseLife();
				dying = false;
				if(lives == 0)
					SetAlive(false);
				else{
					x = spawnX;
					y = spawnY;
					ResetAnimation(LEFT);
					ResetMovement(LEFT);
				}
				
			}
			else
				--curFrame;
		}
	}
}

void PlayerWhite::MoveUp(){
	dirY = -1;
}

void PlayerWhite::MoveDown(){
	dirY = 1;
	if(!isOnSolidGround){
		velY +=gravity;
	}
	else{
		ResetMovement(DOWN);
	}
}

void PlayerWhite::MoveLeft(){
	dirX = -1;
	
	if(frameCount >= frameDelay){
		frameCount = 0;
		//go to start of run animation, if still, or if at end of run animation
		if(animationRow == 8 || (curFrame<=4 && animationRow == 10)){
			animationRow=9;
			curFrame = 3;
		}
		else if((curFrame<=0) && (animationRow == 9)){
			//go to next line of animation
			curFrame = 7;
			animationRow = 10;
		}
		else{
			curFrame--;
		}
	}
}

void PlayerWhite::MoveRight(){
	dirX = 1;
	
	if(frameCount >= frameDelay){
		frameCount = 0;
		//go to start of run animation, if still, or if at end of run animation
		if(animationRow == 8 || (curFrame==3 && animationRow == 1)){
			animationRow=0;
			curFrame = 4;
		}
		else if((curFrame>=7) && (animationRow == 0)){
			//go to next line of animation
			curFrame = 0;
			animationRow = 1;
		}
		else{
			curFrame++;
		}
	}
}

void PlayerWhite::Jump(){
	dirY = 1;
	jump = true;
	isOnSolidGround = false;
	if(mayJumpAgain)
		velY = -jumpSpeed;
	else{
		MoveDown();
	}

	if(isOnSolidGround && !mayJumpAgain){
		ResetMovement(DOWN);
		if(dirX = 1){
			ResetMovement(RIGHT);
			ResetAnimation(RIGHT);
		}
		else if(dirX = -1){
			ResetMovement(LEFT);
			ResetAnimation(LEFT);
		}
		mayJumpAgain = true;
		jump = false;
	}
	else{
		if(dirX == 1){
			animationRow = 5;
			curFrame = 5;
		}
		else if(dirX == -1){
			animationRow = 14;
			curFrame = 2;
		}
	}
	mayJumpAgain = false;
}

bool PlayerWhite::checkMayJumpAgain(){
	if(isOnSolidGround && !jump){
		mayJumpAgain = true;
	}
	else
		mayJumpAgain = false;
	return mayJumpAgain;
}

void PlayerWhite::JumpRight(){
	//animation
	dirX = 1;
	if(frameCount >= frameDelay){
		frameCount = 0;
		if(animationRow != 5){
			animationRow = 5;
			curFrame = 2;
		}
		else if(curFrame == 7){
			willJumpRight = false;
			ResetAnimation(RIGHT);
		}
		else{
		curFrame++;
		}
	}
}
	
void PlayerWhite::JumpLeft(){
	dirX = -1;
	if(frameCount >= frameDelay){
		frameCount = 0;
		if(animationRow != 14){
			animationRow = 14;
			curFrame = 5;
		}
		else if(curFrame == 0){
			willJumpLeft = false;
			ResetAnimation(LEFT);
		}
		else{
		curFrame--;
		}
	}
}		
		
void PlayerWhite::ResetMovement(int direction){
		if(direction == UP){
			dirY = 0;
			velY = 4;
		}
		else if(direction == DOWN){
			dirY = 0;
			velY = 4;
		}
		else if (direction == RIGHT){
			dirX = 0;
			velX = 4;
		}
		else if (direction == LEFT){
			dirX = 0;
			velX = 4;
		}
}

void PlayerWhite::ResetAnimation(int position){
		if(position == UP){
			;
		}
		else if(position == DOWN){
			;
		}
		else if (position == RIGHT){
			curFrame = 0;
			animationRow = 8;
		}
		else if (position == LEFT){
			curFrame = 1;
			animationRow = 8;
		}
}

bool PlayerWhite::CheckCollisions(GameObject *otherObject){
	float oX = otherObject->GetX();
	float oY = otherObject->GetY();

	int obX = otherObject->GetBoundX();
	int obY = otherObject->GetBoundY();

	int oTop = oY-obY;
	int oBottom = oY+obY;
	int oLeft = oX-obX;
	int oRight = oX+obX;

	int top = y-boundY;
	int bottom = y+boundY;
	int left = x-boundX;
	int right = x+boundX;

	if(right > oLeft && 
	   left < oRight &&
	   bottom > oTop &&
	   top < oBottom){

		return  true;
	}else{
		return false;
	}
}

void PlayerWhite::Collided(GameObject *otherObject){
	int objectID = otherObject->GetID();

	float oX = otherObject->GetX();
	float oY = otherObject->GetY();

	int obX = otherObject->GetBoundX();
	int obY = otherObject->GetBoundY();

	int oTop = oY-obY;
	int oBottom = oY+obY;
	int oLeft = oX-obX;
	int oRight = oX+obX;

	int top = y-boundY;
	int bottom = y+boundY;
	int left = x-boundX;
	int right = x+boundX;



	if(objectID == ENEMY){
		if((top<oTop) && (bottom < (oTop+(2*velY)))){
			otherObject->SetAlive(false);
		}
		else
			dieAnimation();
	}
	else if(objectID == FLOOR){
		//Right collision
		if((left < oLeft) && (right < (oLeft + (2*velX)))){
				x-= (right - oLeft); 
		   }
		//Left collision
		if((right > oRight) && (left > (oRight - (2*velX)))){
				x+= (oRight - left);
		   }
		//Bottom collision
		if((top<oTop) && (bottom < (oTop+(2*velY)))){
				y-= (bottom-oTop);
				isOnSolidGround = true;
		   }
		//Top collision
		if((bottom > oBottom) && (top < (oBottom - (2*velY))) && !(((top<oTop) && (bottom < (oTop+(2*velY))))||((left < oLeft) && (right < (oLeft + (2*velX)))))){
				y+=(oBottom-top);
		   }
	}
	else if(objectID == SAW){
		if((std::sqrt(std::pow(right-oX,2)+std::pow(bottom-oY,2)) <= obX) || 
			(std::sqrt(std::pow(left-oX,2)+std::pow(bottom-oY,2)) <= obX) ||
			(std::sqrt(std::pow(right-oX,2)+std::pow(top-oY,2)) <= obX) ||
			(std::sqrt(std::pow(left-oX,2)+std::pow(top-oY,2)) <= obX)){
			dieAnimation();
		}
	}

}

void PlayerWhite::CameraUpdate(float *cameraPosition, float PlayerWhiteX, float PlayerWhiteY, int PlayerWhiteWidth, int PlayerWhiteHeight){
	cameraPosition[0] = (PlayerWhiteX + PlayerWhiteWidth/2)-(WIDTH/2);
	cameraPosition[1] = (PlayerWhiteY + PlayerWhiteWidth/2) - (HEIGHT/2);

	if(cameraPosition[0] < 0)
		cameraPosition[0] = 0;
	if(cameraPosition[1]<0)
		cameraPosition[1] = 0;
}

File: js/playerWhite.js
// PlayerWhite class - the white character

class PlayerWhite extends GameObject {
    constructor() {
        super();
        this.setId(ID.PLAYERWHITE);
        this.setCollidable(true);
        this.setCollideType(COLLIDE_TYPE.WHITE);
        
        this.lives = 3;
        this.score = 0;
        this.animationRow = 0;
        this.gravity = 0.5;
        this.jumpSpeed = -12;
        this.spawnX = 0;
        this.spawnY = 0;
        
        this.dying = false;
        this.jump = false;
        this.mayJumpAgain = true;
        this.isOnSolidGround = false;
        this.jumpFramesLeft = 0;
        this.willJumpLeft = false;
        this.willJumpRight = false;
        this.cameraPosition = [0, 0];
        
        this.moveSpeed = 3;
    }
    
    init(texture, spawnX, spawnY) {
        this.spawnX = spawnX;
        this.spawnY = spawnY;
        this.x = spawnX;
        this.y = spawnY;
        
        this.frameWidth = 32;
        this.frameHeight = 32;
        this.boundX = 16;
        this.boundY = 16;
        
        if (texture) {
            this.sprite = new PIXI.Sprite(texture);
            this.sprite.width = this.frameWidth;
            this.sprite.height = this.frameHeight;
            this.sprite.anchor.set(0.5, 0.5);
            this.sprite.tint = 0xFFFFFF; // White tint
        }
    }
    
    update() {
        // Apply gravity if not on ground or if velocity is upward (jumping)
        if (!this.isOnSolidGround || this.velY < 0) {
            this.velY += this.gravity;
        }
        
        // Update position
        this.x += this.velX;
        this.y += this.velY;
        
        // If jumping upward, no longer on ground
        if (this.velY < 0) {
            this.isOnSolidGround = false;
        }
    }
    
    render() {
        super.render();
    }
    
    moveLeft() {
        this.velX = -this.moveSpeed;
        this.dirX = -1;
    }
    
    moveRight() {
        this.velX = this.moveSpeed;
        this.dirX = 1;
    }
    
    jump() {
        if (this.mayJumpAgain || this.isOnSolidGround) {
            this.velY = this.jumpSpeed;
            this.jump = true;
            this.mayJumpAgain = false;
        }
    }
    
    resetMovement(direction) {
        if (direction === KEYS.LEFT || direction === KEYS.RIGHT) {
            this.velX = 0;
        }
    }
    
    resetAnimation(position) {
        this.curFrame = 0;
        this.frameCount = 0;
    }
    
    checkMayJumpAgain() {
        return this.mayJumpAgain || this.isOnSolidGround;
    }
    
    checkCollisions(otherObject) {
        if (!this.collidable() || !otherObject.collidable()) {
            return false;
        }
        
        // White player collides with BLACK floors ONLY (not white floors)
        const otherType = otherObject.getCollideType();
        if (otherType === COLLIDE_TYPE.WHITE) {
            // console.log("White player checking collision with black object");
            return super.checkCollisions(otherObject);
        }
        
        return false;
    }
    
    collided(otherObject) {
        const otherId = otherObject.getId();
        
        if (otherId === ID.FLOOR && this.velY > 0) {
            // Simple and reliable floor collision
            this.y = otherObject.getY() - otherObject.getBoundY() - this.boundY;
            this.velY = 0;
            this.isOnSolidGround = true;
            this.mayJumpAgain = true;
        } else if (otherId === ID.GOAL) {
            console.log("White player reached goal!");
        }
    }
    
    cameraUpdate(cameraPosition, playerX, playerY, playerWidth, playerHeight) {
        // Simple camera following with smoothing
        const targetX = playerX - WIDTH / 2;
        const targetY = playerY - HEIGHT / 2;
        
        // Smooth camera movement to reduce jittering
        const lerp = 0.1;
        cameraPosition[0] += (targetX - cameraPosition[0]) * lerp;
        cameraPosition[1] += (targetY - cameraPosition[1]) * lerp;
        
    }
    
    // Getters and setters
    getVelX() { return this.velX; }
    getVelY() { return this.velY; }
    setVelX(velX) { this.velX = velX; }
    setVelY(velY) { this.velY = velY; }
    
    getWidth() { return this.frameWidth; }
    getHeight() { return this.frameHeight; }
    
    // Additional getters and setters
    getLives() { return this.lives; }
    getScore() { return this.score; }
    getDirX() { return this.dirX; }
    getDirY() { return this.dirY; }
    
    setDirX(dirX) { this.dirX = dirX; }
    setDirY(dirY) { this.dirY = dirY; }
    
    loseLife() { this.lives--; }
    addPoint() { this.score++; }
}